typing/env.ml

```diff
@@ -1039,22 +1046,45 @@ let rec normalize_path lax env path =
   || (match path with Pident id -> not (Ident.persistent id) | _ -> true) ->
       path
 
-let normalize_path oloc env path =
-  try normalize_path (oloc = None) env path
+let normalize_module_path oloc env path =
+  try normalize_module_path (oloc = None) env path
   with Not_found ->
     match oloc with None -> assert false
     | Some loc ->
-        raise (Error(Missing_module(loc, path, normalize_path true env path)))
+        raise (Error(Missing_module(loc, path,
+                                    normalize_module_path true env path)))
 
 let normalize_path_prefix oloc env path =
   match path with
     Pdot(p, s, pos) ->
-      Pdot(normalize_path oloc env p, s, pos)
+      let p2 = normalize_module_path oloc env p in
+      if p == p2 then path else Pdot(p2, s, pos)
   | Pident _ ->
       path
   | Papply _ ->
       assert false
 
+let is_uident s =
+  assert (s <> "");
+  match s.[0] with
+  | 'A'..'Z' -> true
+  | _ -> false
+
+let normalize_type_path oloc env path =
+  (* Inlined version of Path.constructor_typath *)
```

Comment:
Actually, you're inlining `is_constructor_typath`.
It would be good to add a comment about the structure of those paths: a regular type path is followed by a capitalized constructor name, hence the choice based on the capitalization.